### PR TITLE
Upgrade Gradle to 8.1.1

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

[Gradle 8](https://docs.gradle.org/8.0/release-notes.html) has been released.

Modifications:

- In Gradle 8, dependencies can't be declared in a resolvable configuration (`canBeResolved=true`). `gradle-scripts` used resolvable configurations such as `compileClasspath`, `runtimeClasspath` to inject platform dependencies. https://github.com/line/armeria/blob/20969832640bcc0fa66edc68c531891ab35cea25/gradle/scripts/lib/common-dependencies.gradle#L180-L183
  ```
  * Where: Script '.../gradle/scripts/lib/common-dependencies.gradle'

  * What went wrong: A problem occurred configuring project ':foobar'.
  > Dependencies can not be declared against the `compileClasspath` configuration.
  ```

  `platform` dependencies should be added to non resolvable and non
  configurations. There is no workaround to set dependencies to
  resolvable configurations. So I changed `common-dependencies.gradle`
  to set platform dependencies to the non resolvable and non
  configurations. 

  As a workaround `shadedTestImplementation` is newly added to declare
  dependencies while configuring `shadedTest` task. The original `shadedTestRuntime`
  extends `shadedTestImplementation` and is only used to resolve its dependencies.

  Related discussion: https://discuss.gradle.org/t/problem-with-compileclasspath-after-gradle-8-update/44940

- Gradle 8 also disallows implicit dependencies between two tasks. The solution was quite simple. I added an explicit dependency for the implicit dependency with `.dependsOn(...)` whenever encountering the error.

Result:

You can now use Gradle 8 with `gradle-scripts`.

### CODE REVIEW

File: gradle/scripts/lib/common-dependencies.gradle [hunk 1]
```diff
@@ -15,8 +15,6 @@ buildscript {
     }
 }
 
-def managedDependencyOverrides = [] as Set
```
Comment (ikhoon) [inline gradle/scripts/lib/common-dependencies.gradle:18]: It was an unused code.

File: gradle/scripts/lib/kotlin.gradle [hunk 2]
```diff
@@ -36,9 +36,23 @@ configure(projectsWithFlags('kotlin')) {
         }
 
         afterEvaluate {
-            // A workaround for 'runKtlintCheckOverMainSourceSet' uses this output of task ':generateProto' without
-            // declaring an explicit or implicit dependency.
-            tasks.runKtlintCheckOverMainSourceSet.dependsOn(project.ext.getGenerateSourcesTask())
+
+            // `tasks.withType(KtLintCheckTask)` does not work here.
+            ['Main', 'Test', 'Jmh'].each { name ->
```
Comment (ikhoon) [inline gradle/scripts/lib/kotlin.gradle:41]: `jmh` module injects `runKtlintCheckOverJmhSourceSet` and `compileJmhKotlin` that tasks need explicit dependencies for generated sources from proto or thrift files.

**Comment by trustin** [comment]
Thanks for looking into this potentially very tricky issue, @ikhoon! Let me wait for the build to pass first at the moment.

File: core/build.gradle [hunk 3]
```diff
@@ -173,45 +173,63 @@ if (!rootProject.hasProperty('noWeb')) {
...
+tasks.check.dependsOn testing.suites.testStreaming
 
+// Run the test cases based on reactive-streams-tck only with non-flaky mode
 if (rootProject.findProperty('FLAKY_TESTS') != 'true') {
-    // Run the test cases based on reactive-streams-tck only with non-flaky mode
-    task testNg(type: Test,
-            group: 'Verification',
-            description: 'Runs the TestNG unit tests.',
-            dependsOn: tasks.shadedTestClasses) {
-        useTestNG()
```
Comment (ikhoon) [inline core/build.gradle:201]: Gradle prohibits overriding the test engine with `Test` tasks.
On the other hand, the [JVM Test Suite Plugin](https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html#jvm_test_suite_plugin) supports multiple test engines.
So the test tasks have been migrated to the new test plugin.

File: gradle/scripts/lib/java-shade.gradle [hunk 4]
```diff
@@ -273,23 +286,23 @@ private void configureShadowTask(Project project, ShadowJar task, boolean isMain
  * Finds the dependencies of {@code project} recursively and adds the found dependencies to
  * the configuration named as {@code 'shadedTestRuntime'}.
```
Comment (minwoox) [inline gradle/scripts/lib/java-shade.gradle:287]:
```suggestion
 * the configuration named as {@code 'shadedTestImplementation'}.
```

**Review by minwoox — APPROVED** [review]
Thanks a lot for the upgrade. 🙇

File: spring/boot3-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaSettingsConfigurationTest.java [hunk 5]
```diff
@@ -38,7 +38,7 @@
 @SpringBootTest(classes = TestConfiguration.class)
 @ActiveProfiles({ "local", "settings" })
 @DirtiesContext
-class ArmeriaSettingsConfigurationTest {
+public class ArmeriaSettingsConfigurationTest {
```
Comment (trustin) [inline spring/boot3-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaSettingsConfigurationTest.java:41]: Question: Why was this necessary?

**Review by trustin — APPROVED** [review]
Thank you so much, @ikhoon! This is a labor of love 💘

File: gradle/scripts/lib/common-dependencies.gradle [hunk 6]
```diff
@@ -178,19 +176,8 @@ configure(projectsWithFlags('java')) {
         dependencies {
             configurations.configureEach { configuration ->
                 // Add to resolvable configurations
-                if (configuration.canBeResolved && !configuration.canBeConsumed) {
-                    add(configuration.name, platform(dependencyManagementProject))
-                }
-
-                // Find version overrides in dependency declaration configurations
                 if (!configuration.canBeResolved && !configuration.canBeConsumed) {
```
Comment (jrhee17) [inline gradle/scripts/lib/common-dependencies.gradle:179]: Question) I'm wondering if we want to target resolvable dependencies instead (since these are the configurations which will actually be applied)
```suggestion
                if (configuration.canBeResolved && !configuration.canBeConsumed) {
```

**Review by jrhee17 — APPROVED** [review]
Left a minor question but looks good to me! Thanks @ikhoon 🙇 👍 🚀

File: spring/boot3-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaSettingsConfigurationTest.java (hunk 5 above)
Comment (ikhoon) [reply spring/boot3-autoconfigure/src/test/java/com/linecorp/armeria/spring/ArmeriaSettingsConfigurationTest.java:41]: `ArmeriaSettingsConfigurationTest` uses JUnit 4 which requires a public class.
I don't get why it wasn't detected before.

File: gradle/scripts/lib/common-dependencies.gradle (hunk 6 above)
Comment (ikhoon) [reply gradle/scripts/lib/common-dependencies.gradle:179]: There are three types of configurations.
- Bucket - `implementation`, `api` ...
- Resolvable - `compileClasspath`, `runtimeClasspath`, ...
- Consumable - `apiElement`, ...

We directly added the dependencies to resolvable configurations. In particular, platform dependencies are added to each configuration here. Unfortunately, in Gradle 8, bucket styles are only allowed to add dependencies, and exceptions are raised if the dependencies are added to resolvable ones.

So I changed to add the platform dependencies to the buckets. It should be fine because they will be eventually inherited into resolvable configurations. All resolvable configurations defined in the Gradle Java plugin have at least one bucket configuration.
https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_configurations_graph

The only exception was `shadedTestRuntime` created by `gradle-scripts` that does not have a bucket parent. I added `shadedTestImplementation` as a bucket so that we inject the platform dependencies into it. As `shadedTestRuntime` extends `shadedTestImplementation`, all dependencies added to `shadedTestImplementation` will be visible to `shadedTestRuntime`.

File: gradle/scripts/lib/java.gradle [hunk 7]
```diff
@@ -134,41 +135,49 @@ configure(projectsWithFlags('java')) {
...
-            }
-        } else {
-            if (flakyTests != null) {
-                throw new IllegalArgumentException("flakyTests: $flakyTests (expected: true, false or null)")
-            }
-            useJUnitPlatform()
-        }
-
-        // Use a different JRE for testing if necessary.
-        if (rootProject.ext.buildJdkVersion != rootProject.ext.testJavaVersion) {
```
[the remark this answers is not in the document]
Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:159]: Good idea.

File: gradle/scripts/lib/common-dependencies.gradle (hunk 6 above)
Comment (jrhee17) [reply gradle/scripts/lib/common-dependencies.gradle:179]: I understood the concept while reviewing the PR, but I'm not sure why I left this review 😅  Sorry about that.
In short, I understood that we can't add dependencies to a resolvable configuration 🙇

**Comment by ikhoon** [comment]
Thanks for reviewing. 🙇‍♂️
